app/models/category.rb

```diff
@@ -0,0 +1,17 @@
+class Category < ApplicationRecord
+  has_many :children, class_name: "Category", foreign_key: :parent_id, dependent: :destroy
+  belongs_to :parent, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_category_products, through: :children, dependent: :restrict_with_error, source: :products
+
+  validates :name, presence: true
+  validates :name, uniqueness: true, if: :name?
```

Comment:
please read about allow_blank and allow_nil and see if any can be used in this and the next validation

Reply:
Yes, I can make use of allow_blank here, made appropriate changes